### PR TITLE
mod(es): change definition of ast

### PR DESCRIPTION
**Pull request by Austaras**
Some cases (optional_chaining__optional_super_property and optional_chaining__optional_super_property_class) for parser will be invalid after this change, what should I do? Just delete it?

### CODE REVIEW

**Comment by kdy1** [comment]
Yeah, I think so.

**Comment by dsherret** [comment]
Is there an outline of what changes are taking place and why the changes are taking place somewhere? My browser keeps crashing looking at the diff 😅

**Comment by Austaras** [comment]
> Is there an outline of what changes are taking place and why the changes are taking place? My browser keeps crashing looking at the diff 😅

See [here](https://github.com/swc-project/swc/discussions/3153) or first commit

File: crates/swc_ecma_transforms_base/src/hygiene/usage_analyzer.rs [hunk 1]
```diff
@@ -580,8 +580,14 @@ impl Visit for UsageAnalyzer<'_> {
     fn visit_member_expr(&mut self, e: &MemberExpr) {
         e.obj.visit_with(self);
 
-        if e.computed {
-            e.obj.visit_with(self);
```
Comment (Austaras) [inline crates/swc_ecma_transforms_base/src/hygiene/usage_analyzer.rs:584]: I suppose this should be `e.prop.visit_with` @kdy1 ?

**Comment by Austaras** [comment]
Also, is there a difference between `optimize_property_of_member_expr` and `handle_known_computed_member_expr`? They seems to have identical functionality

File: crates/swc_ecma_transforms_base/src/hygiene/usage_analyzer.rs (hunk 1 above)
Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/hygiene/usage_analyzer.rs:584]: Yes you are right

Comment (Austaras) [reply crates/swc_ecma_transforms_base/src/hygiene/usage_analyzer.rs:584]: Oh, let's spin a separate pr real quick to avoid massive snapshot overwrite...

**Comment by kdy1** [comment]
Hmm... Maybe it's a duplicate logic.
Can you remove one of them?

**Comment by Austaras** [comment]
图片

File: crates/swc_ecma_ast/src/expr.rs [hunk 2]
```diff
@@ -63,6 +64,10 @@ pub enum Expr {
     #[tag("MemberExpression")]
     Member(MemberExpr),
 
+    #[tag("SuperPropExpression")]
+    #[is(name = "super_")]
+    Super(SuperPropExpr),
```
Comment (kdy1) [inline crates/swc_ecma_ast/src/expr.rs:69]: I think `SuperProp` would be better, as `Expr::Super` looks like `super` instead of `super.foo`.

**Review by kdy1 — COMMENTED** [review]
I didn't fully reviewed yet, but I found one concern which will invalidate lots of diffs.

File: crates/swc_ecma_ast/src/expr.rs (hunk 2 above)
Comment (Austaras) [reply crates/swc_ecma_ast/src/expr.rs:69]: If name could remain the same I could push an commit to rename this, and rebase it before update parser change

Comment (Austaras) [reply crates/swc_ecma_ast/src/expr.rs:69]: Well the change is not that big, I could handle it mannually

Comment (kdy1) [reply crates/swc_ecma_ast/src/expr.rs:69]: I disabled auto-merge of parser PR. Thanks!

**Comment by Austaras** [comment]
> Hmm... Maybe it's a duplicate logic. Can you remove one of them?

Seems like remove either one of them would break some tests, I will return to this later

File: node-swc/src/binding.js [hunk 3]
```diff
@@ -5,19 +5,9 @@ const { platform, arch } = process
 
 let nativeBinding = null
 let localFileExisted = false
+let isMusl = false
```
Comment (kdy1) [inline node-swc/src/binding.js]: This seems like a wrong change.

File: crates/swc_ecma_transforms_compat/src/es2015/regenerator/mod.rs [hunk 4]
```diff
@@ -338,8 +338,8 @@ impl Regenerator {
         let ctx = private_ident!("_ctx");
         let mut handler = CaseHandler::new(&ctx);
 
-        f.body
-            .visit_mut_with(&mut FnSentVisitor { ctx: ctx.clone() });
+        // f.body
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2015/regenerator/mod.rs]: Why did you commented this out?

File: crates/swc_ecma_transforms_compat/src/es2020/opt_chaining.rs [hunk 5]
```diff
@@ -554,25 +529,21 @@ impl OptChaining {
...
-                                        left: PatOrExpr::Pat(Box::new(Pat::Ident(
-                                            this_obj.clone().into(),
-                                        ))),
-                                        right: obj.obj.take().expect_expr(),
-                                    })
-                                    .as_obj(),
-                                    prop: obj.prop.take(),
-                                    computed: obj.computed,
-                                })),
+                                Expr::Member(obj @ MemberExpr { .. }) => {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2020/opt_chaining.rs]: ` @ MemberExpr { .. }` seems redundant

**Review by kdy1 — CHANGES_REQUESTED** [review]
Whoa, thank you so much for the hard work.

I used few hours to review this :)

File: crates/swc_ecma_transforms_compat/src/es2020/opt_chaining.rs (hunk 5 above)
Comment (Austaras) [reply crates/swc_ecma_transforms_compat/src/es2020/opt_chaining.rs:532]: It is

**Comment by Austaras** [comment]
Thank you for take your time to review this

**Review by kdy1 — APPROVED** [review]

**Pull request merged**